File: JAVA/KadaneAlgorithm.java
import java.util.Arrays;

public class KadaneAlgorithm {

    // Function to find maximum subarray sum
    public static int maxSubArray(int[] nums) {
        int maxSoFar = nums[0];
        int maxEndingHere = nums[0];

        for (int i = 1; i < nums.length; i++) {
            // Either extend the current subarray or start a new one from current element
            maxEndingHere = Math.max(nums[i], maxEndingHere + nums[i]);
            // Update maxSoFar if needed
            maxSoFar = Math.max(maxSoFar, maxEndingHere);
        }

        return maxSoFar;
    }

    public static void main(String[] args) {
        int[] nums1 = {2, 3, 5, -2, 7, -4};
        int[] nums2 = {-2, -3, -7, -2, -10, -4};
        int[] nums3 = {-1, 2, 3, -1, 2, -6, 5};

        System.out.println(maxSubArray(nums1)); // 15
        System.out.println(maxSubArray(nums2)); // -2
        System.out.println(maxSubArray(nums3)); // 5
    }
}

File: Python/KadaneAlgorithm.py
def max_subarray(nums):
    max_so_far = nums[0]
    max_ending_here = nums[0]

    for num in nums[1:]:
        # Either extend the current subarray or start new from current element
        max_ending_here = max(num, max_ending_here + num)
        # Update max_so_far if needed
        max_so_far = max(max_so_far, max_ending_here)

    return max_so_far

# Example usage
nums1 = [2, 3, 5, -2, 7, -4]
nums2 = [-2, -3, -7, -2, -10, -4]
nums3 = [-1, 2, 3, -1, 2, -6, 5]

print(max_subarray(nums1))  # 15
print(max_subarray(nums2))  # -2
print(max_subarray(nums3))  # 5
